From 406d84bfd8e7ff57b86b069ffcce3a79884218fd Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Sun, 13 Nov 2022 18:06:08 +0100
Subject: [PATCH 182/196] Phase out Taproot and SIGHASH_LOTUS (#289)

1. Reject transactions with Taproot outputs from being added to the mempool. We do this by adding TxHasPayToTaproot, which is checked in `MemPoolAccept::PreChecks`.
2. Add `SCRIPT_DISABLE_TAPROOT_SIGHASH_LOTUS` script flag, which makes the interpreter return an error when the user wants to spend a taproot script, or use the SIGHASH_LOTUS sighash type.
3. Add  `SCRIPT_DISABLE_TAPROOT_SIGHASH_LOTUS` to `STANDARD_SCRIPT_VERIFY_FLAGS`.
4. Before Numbers activates, remove `SCRIPT_DISABLE_TAPROOT_SIGHASH_LOTUS` from required script flags so we don't split the network.
5. After Numbers activated, transactions with Taproot output are now violating consensus.

Co-authored-by: Shammah Chancellor <shammah.chancellor@gmail.com>
---
 src/policy/policy.h                           |   2 +-
 src/script/interpreter.cpp                    |   7 +
 src/script/script_error.cpp                   |   2 +
 src/script/script_error.h                     |   3 +
 src/script/script_flags.h                     |   4 +
 src/script/taproot.cpp                        |  10 ++
 src/script/taproot.h                          |   4 +
 src/test/scriptflags.cpp                      |   1 +
 src/test/taproot_tests.cpp                    |   3 +-
 src/test/transaction_tests.cpp                |   8 -
 src/validation.cpp                            |  22 +++
 .../functional/logos_feature_lotus_sighash.py |   4 +
 .../logos_feature_taproot_key_spend.py        |   4 +
 .../logos_feature_taproot_phaseout.py         | 159 ++++++++++++++++++
 .../logos_feature_taproot_script_path.py      |  15 +-
 test/functional/rpc_signrawtransaction.py     |   6 -
 16 files changed, 227 insertions(+), 27 deletions(-)
 create mode 100644 test/functional/logos_feature_taproot_phaseout.py

diff --git a/src/policy/policy.h b/src/policy/policy.h
index 27c541bfe..d66af17b0 100644
--- a/src/policy/policy.h
+++ b/src/policy/policy.h
@@ -77,7 +77,7 @@ static constexpr uint32_t MANDATORY_SCRIPT_VERIFY_FLAGS =
  */
 static constexpr uint32_t STANDARD_SCRIPT_VERIFY_FLAGS =
     MANDATORY_SCRIPT_VERIFY_FLAGS | SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_NOPS |
-    SCRIPT_VERIFY_INPUT_SIGCHECKS;
+    SCRIPT_VERIFY_INPUT_SIGCHECKS | SCRIPT_DISABLE_TAPROOT_SIGHASH_LOTUS;
 
 /**
  * For convenience, standard but not mandatory verify flags.
diff --git a/src/script/interpreter.cpp b/src/script/interpreter.cpp
index 1782ad8c9..d1abc9e8d 100644
--- a/src/script/interpreter.cpp
+++ b/src/script/interpreter.cpp
@@ -1869,6 +1869,10 @@ bool SignatureHash(uint256 &sighashOut,
                                    sigHashType, amount, cache);
     } else if (sigHashType.hasLotus()) {
         assert(cache);
+        if (flags & SCRIPT_DISABLE_TAPROOT_SIGHASH_LOTUS) {
+            // SIGHASH_LOTUS phaseout
+            return false;
+        }
         return SignatureHashLotus(sighashOut, execdata, txTo, nIn, sigHashType,
                                    *cache);
     } else {
@@ -2073,6 +2077,9 @@ static bool VerifyTaprootSpend(std::vector<valtype> stack,
                                const BaseSignatureChecker &checker,
                                ScriptExecutionMetrics &metrics,
                                ScriptError *serror) {
+    if (flags & SCRIPT_DISABLE_TAPROOT_SIGHASH_LOTUS) {
+        return set_error(serror, ScriptError::TAPROOT_PHASEOUT);
+    }
     if (!IsPayToTaproot(script_pubkey)) {
         return set_error(serror, ScriptError::SCRIPTTYPE_MALFORMED_SCRIPT);
     }
diff --git a/src/script/script_error.cpp b/src/script/script_error.cpp
index a6560959a..5f1cf971d 100644
--- a/src/script/script_error.cpp
+++ b/src/script/script_error.cpp
@@ -128,6 +128,8 @@ std::string ScriptErrorString(const ScriptError serror) {
             return "Taproot control does not verify the script";
         case ScriptError::TAPROOT_LEAF_VERSION_NOT_SUPPORTED:
             return "Unsupported taproot script leaf version";
+        case ScriptError::TAPROOT_PHASEOUT:
+            return "Taproot is being phased out";
         case ScriptError::UNKNOWN:
         case ScriptError::ERROR_COUNT:
         default:
diff --git a/src/script/script_error.h b/src/script/script_error.h
index 009be15f7..a7aedb093 100644
--- a/src/script/script_error.h
+++ b/src/script/script_error.h
@@ -108,6 +108,9 @@ enum class ScriptError {
     /* Taproot leaf version not supported */
     TAPROOT_LEAF_VERSION_NOT_SUPPORTED,
 
+    /* Taproot is being phased out */
+    TAPROOT_PHASEOUT,
+
     ERROR_COUNT,
 };
 
diff --git a/src/script/script_flags.h b/src/script/script_flags.h
index 24a9ad5e5..493fb9140 100644
--- a/src/script/script_flags.h
+++ b/src/script/script_flags.h
@@ -16,6 +16,10 @@ enum {
     // Only used for checking signatures, not during script execution.
     SCRIPT_TAPROOT_KEY_SPEND_PATH = (1U << 0),
 
+    // Disable the execution of Taproot scripts, as well as SIGHASH_LOTUS
+    // sighash types.
+    SCRIPT_DISABLE_TAPROOT_SIGHASH_LOTUS = (1U << 1),
+
     // Discourage use of NOPs reserved for upgrades (NOP1-10)
     //
     // Provided so that nodes can avoid accepting or mining transactions
diff --git a/src/script/taproot.cpp b/src/script/taproot.cpp
index f7319b1e6..8965c1d86 100644
--- a/src/script/taproot.cpp
+++ b/src/script/taproot.cpp
@@ -4,6 +4,7 @@
 
 #include <hash.h>
 #include <pubkey.h>
+#include <primitives/transaction.h>
 #include <script/script.h>
 #include <script/taproot.h>
 
@@ -84,3 +85,12 @@ bool IsPayToTaproot(const CScript &script) {
     return script.size() == TAPROOT_SIZE_WITH_STATE &&
            script[TAPROOT_SIZE_WITHOUT_STATE] == CSHA256::OUTPUT_SIZE;
 }
+
+bool TxHasPayToTaproot(const CTransaction &tx) {
+    for (CTxOut output : tx.vout) {
+        if (IsPayToTaproot(output.scriptPubKey)) {
+            return true;
+        }
+    }
+    return false;
+}
diff --git a/src/script/taproot.h b/src/script/taproot.h
index f4ddfd5d6..1ab6c2452 100644
--- a/src/script/taproot.h
+++ b/src/script/taproot.h
@@ -10,6 +10,8 @@
 
 typedef std::vector<uint8_t> valtype;
 
+class CTransaction;
+
 static constexpr uint8_t TAPROOT_LEAF_MASK = 0xfe;
 static constexpr uint8_t TAPROOT_LEAF_TAPSCRIPT = 0xc0;
 static constexpr size_t TAPROOT_CONTROL_BASE_SIZE = 33;
@@ -57,4 +59,6 @@ bool VerifyTaprootCommitment(uint256 &tapleaf_hash,
  */
 bool IsPayToTaproot(const CScript &script);
 
+bool TxHasPayToTaproot(const CTransaction &tx);
+
 #endif // BITCOIN_SCRIPT_TAPROOT_H
diff --git a/src/test/scriptflags.cpp b/src/test/scriptflags.cpp
index cde752a5f..ec0230aae 100644
--- a/src/test/scriptflags.cpp
+++ b/src/test/scriptflags.cpp
@@ -15,6 +15,7 @@
 
 static std::map<std::string, uint32_t> mapFlagNames = {
     {"NONE", SCRIPT_VERIFY_NONE},
+    {"DISABLE_TAPROOT_SIGHASH_LOTUS", SCRIPT_DISABLE_TAPROOT_SIGHASH_LOTUS},
     {"DISCOURAGE_UPGRADABLE_NOPS", SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_NOPS},
     {"CLEANSTACK", SCRIPT_VERIFY_CLEANSTACK},
     {"MINIMALIF", SCRIPT_VERIFY_MINIMALIF},
diff --git a/src/test/taproot_tests.cpp b/src/test/taproot_tests.cpp
index 523e16e22..360dc6f5f 100644
--- a/src/test/taproot_tests.cpp
+++ b/src/test/taproot_tests.cpp
@@ -284,7 +284,8 @@ BOOST_AUTO_TEST_CASE(is_p2tr) {
 }
 
 std::array<uint32_t, 2> flagset{
-    {SCRIPT_ENABLE_SIGHASH_FORKID, STANDARD_SCRIPT_VERIFY_FLAGS}};
+    {SCRIPT_ENABLE_SIGHASH_FORKID,
+     STANDARD_SCRIPT_VERIFY_FLAGS & ~SCRIPT_DISABLE_TAPROOT_SIGHASH_LOTUS}};
 
 struct TestUtxo {
     CScript script_sig;
diff --git a/src/test/transaction_tests.cpp b/src/test/transaction_tests.cpp
index 94a898a26..07e9e9fa6 100644
--- a/src/test/transaction_tests.cpp
+++ b/src/test/transaction_tests.cpp
@@ -406,14 +406,6 @@ BOOST_AUTO_TEST_CASE(test_big_transaction) {
     sigHashes.emplace_back(SIGHASH_SINGLE | SIGHASH_FORKID |
                            SIGHASH_ANYONECANPAY);
     sigHashes.emplace_back(SIGHASH_ALL | SIGHASH_FORKID | SIGHASH_ANYONECANPAY);
-    sigHashes.emplace_back(SIGHASH_NONE | SIGHASH_LOTUS);
-    sigHashes.emplace_back(SIGHASH_SINGLE | SIGHASH_LOTUS);
-    sigHashes.emplace_back(SIGHASH_ALL | SIGHASH_LOTUS);
-    sigHashes.emplace_back(SIGHASH_NONE | SIGHASH_LOTUS |
-                           SIGHASH_ANYONECANPAY);
-    sigHashes.emplace_back(SIGHASH_SINGLE | SIGHASH_LOTUS |
-                           SIGHASH_ANYONECANPAY);
-    sigHashes.emplace_back(SIGHASH_ALL | SIGHASH_LOTUS | SIGHASH_ANYONECANPAY);
 
     CMutableTransaction mtx;
     mtx.nVersion = 1;
diff --git a/src/validation.cpp b/src/validation.cpp
index 4c00bf6b0..e274bf8d3 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -39,6 +39,7 @@
 #include <script/script.h>
 #include <script/scriptcache.h>
 #include <script/sigcache.h>
+#include <script/taproot.h>
 #include <shutdown.h>
 #include <timedata.h>
 #include <tinyformat.h>
@@ -463,6 +464,11 @@ bool MemPoolAccept::PreChecks(ATMPArgs &args, Workspace &ws) {
         return state.Invalid(TxValidationResult::TX_NOT_STANDARD, reason);
     }
 
+    if (fRequireStandardPolicy && TxHasPayToTaproot(tx)) {
+        return state.Invalid(TxValidationResult::TX_NOT_STANDARD,
+                             "bad-taproot-phased-out");
+    }
+
     // Only accept nLockTime-using transactions that can be mined in the next
     // block; we don't want our mempool filled up with transactions that can't
     // be mined yet.
@@ -1591,6 +1597,10 @@ static uint32_t GetNextBlockScriptFlags(const Consensus::Params &params,
         flags |= SCRIPT_ENABLE_REPLAY_PROTECTION;
     }
 
+    if (IsNumbersEnabled(params, pindex)) {
+        flags |= SCRIPT_DISABLE_TAPROOT_SIGHASH_LOTUS;
+    }
+
     return flags;
 }
 
@@ -1722,6 +1732,11 @@ bool CChainState::ConnectBlock(const CBlock &block, BlockValidationState &state,
         // the mempool, otherwise we might mine invalid blocks.
         assert(fRequireStandardPolicy);
         extraFlags = STANDARD_SCRIPT_VERIFY_FLAGS;
+
+        // Before the Numbers upgrade, we don't deny Taproot or SIGHASH_LOTUS
+        if (!IsNumbersEnabled(consensusParams, pindex->pprev)) {
+            extraFlags &= ~SCRIPT_DISABLE_TAPROOT_SIGHASH_LOTUS;
+        }
     }
 
     const uint32_t flags =
@@ -1808,6 +1823,13 @@ bool CChainState::ConnectBlock(const CBlock &block, BlockValidationState &state,
                 return state.Invalid(BlockValidationResult::BLOCK_CONSENSUS,
                                      reason);
             }
+            // Taproot is phased out after the Numbers update
+            if (IsNumbersEnabled(consensusParams, pindex->pprev)) {
+                if (TxHasPayToTaproot(tx)) {
+                    return state.Invalid(BlockValidationResult::BLOCK_CONSENSUS,
+                                         "bad-taproot-phased-out");
+                }
+            }
         }
 
         if (!MoneyRange(nFees)) {
diff --git a/test/functional/logos_feature_lotus_sighash.py b/test/functional/logos_feature_lotus_sighash.py
index 8aaab1d91..11caa87f9 100755
--- a/test/functional/logos_feature_lotus_sighash.py
+++ b/test/functional/logos_feature_lotus_sighash.py
@@ -222,12 +222,16 @@ TESTCASES = [
 ]
 
 ACTIVATION_TIME = 2000000000
+NUMBERS_ACTIVATION_TIME = 2010000000
 
 class Bip341Sighash(BitcoinTestFramework):
     def set_test_params(self):
         self.num_nodes = 1
         self.setup_clean_chain = True
         self.extra_args = [['-whitelist=noban@127.0.0.1',
+                            '-acceptnonstdtxn=1',
+                            '-allownonstdtxnconsensus=1',
+                            f'-numbersactivationtime={NUMBERS_ACTIVATION_TIME}',
                             f'-replayprotectionactivationtime={ACTIVATION_TIME}']]
 
     BASE_TYPES = {
diff --git a/test/functional/logos_feature_taproot_key_spend.py b/test/functional/logos_feature_taproot_key_spend.py
index b7e84f325..9b6ecfd0e 100644
--- a/test/functional/logos_feature_taproot_key_spend.py
+++ b/test/functional/logos_feature_taproot_key_spend.py
@@ -133,12 +133,16 @@ TESTCASES = [
 ]
 
 ACTIVATION_TIME = 2000000000
+NUMBERS_ACTIVATION_TIME = 2010000000
 
 class TaprootKeySpendTest(BitcoinTestFramework):
     def set_test_params(self):
         self.num_nodes = 1
         self.setup_clean_chain = True
         self.extra_args = [['-whitelist=noban@127.0.0.1',
+                            '-acceptnonstdtxn=1',
+                            '-allownonstdtxnconsensus=1',
+                            f'-numbersactivationtime={NUMBERS_ACTIVATION_TIME}',
                             f'-replayprotectionactivationtime={ACTIVATION_TIME}']]
 
     def run_test(self):
diff --git a/test/functional/logos_feature_taproot_phaseout.py b/test/functional/logos_feature_taproot_phaseout.py
new file mode 100644
index 000000000..6913dd7d8
--- /dev/null
+++ b/test/functional/logos_feature_taproot_phaseout.py
@@ -0,0 +1,159 @@
+#!/usr/bin/env python3
+# Copyright (c) 2021 The Logos Foundation
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""
+Test that we successfully phased out taproot during the Numbers update
+"""
+
+from test_framework.key import ECKey
+from test_framework.blocktools import (
+    SUBSIDY,
+    create_coinbase,
+    create_block,
+    prepare_block,
+)
+from test_framework.messages import (
+    COutPoint,
+    CTransaction,
+    CTxIn,
+    CTxOut,
+)
+from test_framework.p2p import (
+    P2PDataStore,
+)
+from test_framework.script import (
+    CScript,
+    OP_1,
+    OP_EQUAL,
+    OP_HASH160,
+    OP_SCRIPTTYPE,
+    SIGHASH_ALL,
+    SIGHASH_LOTUS,
+    SignatureHashLotus,
+)
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.txtools import pad_tx
+
+
+NUMBERS_ACTIVATION_TIME = 2000000000
+REPLAY_ACTIVATION_TIME = 2010000000
+
+class TaprootKeySpendTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 1
+        self.setup_clean_chain = True
+        self.extra_args = [['-whitelist=noban@127.0.0.1',
+                            f'-numbersactivationtime={NUMBERS_ACTIVATION_TIME}',
+                            f'-replayprotectionactivationtime={REPLAY_ACTIVATION_TIME}']]
+
+    def run_test(self):
+        node = self.nodes[0]
+        peer = node.add_p2p_connection(P2PDataStore())
+
+        value = int(SUBSIDY * 1_000_000)
+        fee = 10_000
+
+        private_key = ECKey()
+        private_key.generate()
+        public_key = private_key.get_pubkey()
+
+        taproot_script = CScript([OP_SCRIPTTYPE, OP_1, public_key.get_bytes()])
+
+        anyonecanspend_address = node.decodescript('51')['p2sh']
+        burn_address = node.decodescript('00')['p2sh']
+        node.generatetoaddress(1, anyonecanspend_address)
+        node.generatetoaddress(100, burn_address)
+
+        # Build taproot funding transaction
+        block_hash = node.getblockhash(1)
+        coin = int(node.getblock(block_hash)['tx'][0], 16)
+        tx_fund_taproot = CTransaction()
+        tx_fund_taproot.vin.append(CTxIn(COutPoint(coin, 1), CScript([b'\x51'])))
+        tx_fund_taproot.vout.append(CTxOut(value - fee, taproot_script))
+        pad_tx(tx_fund_taproot)
+
+        # Can't add taproot to mempool: They're henceforth non-standard
+        peer.send_txs_and_test([tx_fund_taproot], node, success=False, reject_reason='bad-taproot-phased-out')
+
+        # However, before Numbers, we can still mine txs with taproot output
+        template = node.getblocktemplate()
+        coinbase = create_coinbase(template['height'], pubkey=b'\x03'*33)
+        block1 = create_block(coinbase=coinbase, tmpl=template)
+        block1.vtx.append(tx_fund_taproot)
+        prepare_block(block1)
+        peer.send_blocks_and_test([block1], node, timeout=5)
+
+        # Make tx spending taproot
+        tx_spend_taproot = CTransaction()
+        tx_spend_taproot.vin.append(CTxIn(COutPoint(tx_fund_taproot.txid, 0), CScript()))
+        tx_spend_taproot.vout.append(CTxOut(value - fee * 2, CScript([OP_HASH160, bytes(20), OP_EQUAL])))
+        sighash_type = SIGHASH_ALL | SIGHASH_LOTUS
+        sighash = SignatureHashLotus(
+            tx_to=tx_spend_taproot,
+            spent_utxos=[tx_fund_taproot.vout[0]],
+            sig_hash_type=sighash_type,
+            input_index=0,
+        )
+        signature = private_key.sign_schnorr(sighash)
+        signature += bytes([sighash_type])
+        tx_spend_taproot.vin[0].scriptSig = CScript([signature])
+
+        # We can't add a taproot spend to mempool either
+        peer.send_txs_and_test([tx_spend_taproot], node, success=False, reject_reason='non-mandatory-script-verify-flag (Taproot is being phased out)')
+
+        # However, before Numbers, we can still mine taproot spend txs
+        template = node.getblocktemplate()
+        coinbase = create_coinbase(template['height'], pubkey=b'\x03'*33)
+        block2 = create_block(coinbase=coinbase, tmpl=template)
+        block2.vtx.append(tx_spend_taproot)
+        prepare_block(block2)
+        peer.send_blocks_and_test([block2], node, timeout=5)
+
+        # Undo blocks with the 2 taproot txs
+        node.invalidateblock(block2.hash)
+        node.invalidateblock(block1.hash)
+
+        # Activate Numbers upgrade
+        node.setmocktime(NUMBERS_ACTIVATION_TIME)
+        node.generatetoaddress(6, burn_address)
+
+        # Adding taproot output to mempool still fails
+        peer.send_txs_and_test([tx_fund_taproot], node, success=False, reject_reason='bad-taproot-phased-out')
+
+        # But now taproot outputs can't even be mined in blocks anymore
+        template = node.getblocktemplate()
+        coinbase = create_coinbase(template['height'], pubkey=b'\x03'*33)
+        block3 = create_block(coinbase=coinbase, tmpl=template)
+        block3.vtx.append(tx_fund_taproot)
+        prepare_block(block3)
+        peer.send_blocks_and_test([block3], node, success=False, reject_reason='bad-taproot-phased-out')
+
+        # Now we roll back Numbers
+        node.invalidateblock(node.getbestblockhash())
+
+        # Adding taproot output to mempool still fails
+        peer.send_txs_and_test([tx_fund_taproot], node, success=False, reject_reason='bad-taproot-phased-out')
+
+        # Since we rolled back Numbers, we can mine the taproot output again
+        template = node.getblocktemplate()
+        coinbase = create_coinbase(template['height'], pubkey=b'\x03'*33)
+        block4 = create_block(coinbase=coinbase, tmpl=template)
+        block4.vtx.append(tx_fund_taproot)
+        prepare_block(block4)
+        peer.send_blocks_and_test([block4], node, timeout=5)
+
+        # Adding taproot spend to mempool still fails
+        peer.send_txs_and_test([tx_spend_taproot], node, success=False, reject_reason='non-mandatory-script-verify-flag (Taproot is being phased out)')
+
+        # Numbers is now activated again, so we can't spend the taproot output even in blocks
+        template = node.getblocktemplate()
+        coinbase = create_coinbase(template['height'], pubkey=b'\x03'*33)
+        block5 = create_block(coinbase=coinbase, tmpl=template)
+        block5.vtx.append(tx_spend_taproot)
+        prepare_block(block5)
+        peer.send_blocks_and_test([block5], node, success=False, reject_reason='blk-bad-inputs')
+
+
+if __name__ == '__main__':
+    TaprootKeySpendTest().main()
diff --git a/test/functional/logos_feature_taproot_script_path.py b/test/functional/logos_feature_taproot_script_path.py
index ae4f779ce..6e4763056 100644
--- a/test/functional/logos_feature_taproot_script_path.py
+++ b/test/functional/logos_feature_taproot_script_path.py
@@ -104,17 +104,6 @@ TX_CASES = [
              0, [SIGHASH_ALL | SIGHASH_LOTUS] * 23),
         ],
     ),
-    dict(
-        outputs=8,
-        inputs=[
-            (dict(script_inputs=[],
-                  script=[PUBLIC_KEYS[4]] + [OP_2DUP, OP_CHECKSIGVERIFY] * 40 + [OP_CHECKSIG],
-                  keys=[PRIVATE_KEYS[4]],
-                  schnorr=True),
-             0, [SIGHASH_ALL | SIGHASH_LOTUS])
-        ],
-        error='Input SigChecks limit exceeded',
-    ),
     dict(
         outputs=6,
         inputs=[
@@ -362,6 +351,7 @@ TX_CASES = [
 ]
 
 ACTIVATION_TIME = 2000000000
+NUMBERS_ACTIVATION_TIME = 2010000000
 
 def ser_script(script: bytes) -> bytes:
     if len(script) < 0xfd:
@@ -396,6 +386,9 @@ class TaprootScriptSpendTest(BitcoinTestFramework):
         self.num_nodes = 1
         self.setup_clean_chain = True
         self.extra_args = [['-whitelist=noban@127.0.0.1',
+                            '-acceptnonstdtxn=1',
+                            '-allownonstdtxnconsensus=1',
+                            f'-numbersactivationtime={NUMBERS_ACTIVATION_TIME}',
                             f'-replayprotectionactivationtime={ACTIVATION_TIME}']]
 
     def run_test(self):
diff --git a/test/functional/rpc_signrawtransaction.py b/test/functional/rpc_signrawtransaction.py
index 7c60e732d..7f1ceab7b 100755
--- a/test/functional/rpc_signrawtransaction.py
+++ b/test/functional/rpc_signrawtransaction.py
@@ -180,12 +180,6 @@ class SignRawTransactionsTest(BitcoinTestFramework):
             "ALL|FORKID|ANYONECANPAY",
             "NONE|FORKID|ANYONECANPAY",
             "SINGLE|FORKID|ANYONECANPAY",
-            "ALL|LOTUS",
-            "NONE|LOTUS",
-            "SINGLE|LOTUS",
-            "ALL|LOTUS|ANYONECANPAY",
-            "NONE|LOTUS|ANYONECANPAY",
-            "SINGLE|LOTUS|ANYONECANPAY",
         ]
         no_forkid_sighashes = [
             "ALL",
-- 
2.39.5 (Apple Git-154)

